src/Controller/ControllerFactory.php

```diff
@@ -73,9 +98,49 @@ public function invoke($controller): ResponseInterface
         if ($result instanceof ResponseInterface) {
             return $result;
         }
-
         $action = $controller->getAction();
-        $args = array_values($controller->getRequest()->getParam('pass'));
+
+        $args = [];
+        $reflection = new ReflectionFunction($action);
+        $passed = array_values((array)$controller->getRequest()->getParam('pass'));
+        foreach ($reflection->getParameters() as $i => $parameter) {
+            $position = $parameter->getPosition();
+
+            // If there is no type we can't look in the container
+            // assume the parameter is a passed param
+            $type = $parameter->getType();
+            if (!$type) {
+                $args[$position] = array_shift($passed);
+                continue;
+            }
+            $typeName = $type->getName();
+            if (substr($typeName, 0, 1) === '?') {
+                $typeName = substr($typeName, 1);
+            }
+
+            // Primitive types are passed args as they can't be looked up in the container.
+            if (in_array($typeName, ['float', 'int', 'string', 'bool'], true)) {
+                if (count($passed) || !$type->allowsNull()) {
+                    $args[$position] = array_shift($passed);
```

Comment:
Passed args from URL would be always string, so wouldn't we need to typecast the arg if the param is typehinted with any primitive type besides `string`?

Reply:
Yeah we would. Historically we've not done that typecasting. We could add that now, but we could run into complexity around handling invalid data. I'll change this to only look for strings instead. Adding other primitive types to passed arguments can be a separate discussion.

Reply:
Yeah the code checks for all primitive types hence my comment. If we aren't going to typecast then we need to assume that the method param can only have string type or none specified.